rfc/0001-policy-server-custom-resource.md

```diff
@@ -0,0 +1,426 @@
+> - What is the expected outcome?
+
+> Describe the problem you are trying to solve, and its constraints, without coupling them too closely to the solution you have in mind. If this RFC is not accepted, the motivation can be used to develop alternative solutions.
+
+Currently the Kubewarden controller deploys a single Deployment of Policy Server.
+All the `ClusterAdmissionPolicies` are then evaluated by the Policy Server Pods
+that belong to this Deployment.
+The configuration of this Policy Server Deployment is kept inside of a ConfigMap
+called `policy-server`.
+
+Right now, the kubewarden-controller is monitoring the ClusterAdmissionPolicies
+resources. Changes against these resources trigger an update of the `policy-server`
+ConfigMap, and then a restart of the Policy Server Deployment.
+
+The `policy-server` ConfigMap contains a series of configuration options related
+with the Deployment (e.g.: the image to use, the size of the replica,...) and
+a snapshot of the `policies.yml` file. The latter one is the file that contains
+the list of policies to load, where to pull them, their configuration options,...
+
+Currently, user-made changes to the `policy-server` ConfigMap (e.g. manual
+edit via `kubectl`, or `helm upgrade`) **do not** lead to the Policy Server Deployment
+to be updated.
+Moreover, when changing the `policy-server` ConfigMap, the user
+must be really careful to not alter the contents of the `policies.yml` key. That
+can result in catastrophic outcome (all policies being dropped by the Policy Server,
+leading to massive failures to all the admission reviews, leading to potential
+denial of service against the Kubernetes cluster).
+
+> **Problem #1:** provide a better UX to update the tuning
+> options of the Policy Server Deployment. This must be easy and safe.
+
+Thinking long term, there's a scalability and reliability issue with our current
+architecture. All the policies are loaded by each Policy Server process. Inherently,
+all the admission reviews are sent to the single deployment.
+
+While the current setup is already HA and can be scaled by increasing the number
+of replicas, there are still some hard limits and not-so-uncommon scenarios
+that could lead to a service disruption:
+
+  * Defining too many ClusterAdmissionPolicies can lead to Policy Server processes
+    consuming a lot of memory and even hit resource limits
+  * A "noisy" tenant/namespace or a frequently used policy can slow down policy
+    evaluation and potentially bring down a Policy Server instance
+
+> **Problem #2:** provide a way to partition policies. Allow the operator to create
+> dedicated Policy Server Deployments, handling only a chosen set of policies.
+
+This RFC aims to solve both problems.
+
+The proposed solution is to:
+
+  * Introduce a new Custom Resource Definition describing a PolicyServer deployment
+  * Create a parent-child relationship between a PolicyServer CR and the ClusterAdmissionPolicies
```

Comment:
```suggestion
  * Create a parent-child relationship between a PolicyServer CR and the ClusterAdmissionPolicies
```

After a second thought, I think is better if we talk about "binding" rather than "parent-child" relationship.

Reply:
I've resolved that using the `binding` term